### PR TITLE
clusterversion: prevent upgrades from development versions

### PR DESCRIPTION
**Pull request by dt**
This change defines a new "unstableVersionsAbove" point on the cluster
version  line, above which any cluster versions are considered unstable
development-only versions which are still subject to change.

Performing an upgrade to a version while it is still unstable leaves a
cluster in a state where it persists a version that claims it has done
that upgrade and all prior, however those upgrades are still subject to
change by nature of being unstable. If it subsequently upgraded to a
stable version, this could result in subtle and nearly impossible to
detect issues, as being at or above a particular version is used to
assume that all subsequent version upgrades _as released_ were run; on a
cluster that ran an earlier iteration of an upgrade this does not hold.

Thus to prevent clusters which upgrade to development versions from
subsequently upgrading to a stable version, we offset all development
versions -- those above the unstableVersionsAbove point -- into the far
future by adding one million to their major version e.g. v22.x-y becomes
1000022.x-y. This means an attempt to subsequently "upgrade" to a stable
version -- such as v22.2 -- will look like a downgrade and be forbidden.

On the release branch, prior to starting to publish upgradable releases,
the unstableVersionsAbove value should be set to invalidVersionKey to
reflect that all version upgrades in that release branch are now
considered to be stable, meaning they must be treated as immutable and
append-only.

Release note (ops change): clusters that are upgraded to an alpha or
other manual build from the development branch will not be able to be
subsequently upgraded to a release build.

Release justification: high-priority change to existing functionality,
to allow releasing alphas with known version upgrade bugs while ensuring
they do not subsequently upgrade into stable version but silently
corrupted clusters.

### CODE REVIEW

**Comment by dt** [comment]
I went back and forth on what the flow should be on release branches: should every addition of a version -- of which there should be few / none after the mint -- adjust the unstable-above pointer to end-of-list? I didn't want to to be automatic, at least not on master, as that'd defeat the point. Should I condition on isReleaseBranch? Should release team just mint a new version for each beta/rc and be in charge of moving the pointer when they do so, e.g. so a random build from the release branch between RCs would call any new versions since the last unstable? What about the fact that we just forgot to set isReleaseBranch on a release branch for months?

I think I found something I'm happy with in the most recent revisions here though, doing away with the mistake-prone isReleaseBranch while I'm at it: 
a) to serve the purpose of preventing version backports post-mint, I switched `isReleaseBranch` bool to `finalVersion` version key that, if set, must match binary version _and added a check_ that if the binary version has no Internal - i.e. if it looks like a minted final - that final is indeed set (to prevent the forgetting part). 
b) the designation of unstable versions is just a pointer into the line, and it can be unset to disable.

Thus, I believe the flow now would be:
* master has finalVersion=invalid, unstableVersionsAbove=VFinalPrevRelease
* cut release branch
* on release branch ahead of beta 1, unstableVersionsAbove=invalid
* betas, rcs, etc until we're ready to mint VFinal
* on master we mint the final version
  * commit 1: mint VFinal, set finalVersion= VFinal
  * commit 2: mint VNextReleaseStart, set finalVersion=invalid and set unstableVersionsAbove=VFinal
* backport to release branch mint only grabs commit 1

**Review by celiala — APPROVED** [review]
Thanks for putting this safeguard into place. 

This makes sense to me -- :lgtm: 

> Reviewed 2 of 4 files at r3, 12 of 12 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/86345#-:-N9rmuWH42XkB9xmvxoO:b1l96ez)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ajwerner, @bdarnell, @dt, and @stevendanna)

___
*[`pkg/clusterversion/cockroach_versions.go` line 536 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/86345#-N9rmaOe1JlaroJtGEZb:-N9rmaOe1JlaroJtGEZc:b-g2a4cz) ([raw file](https://github.com/cockroachdb/cockroach/blob/5a511c2975faedf120ebdc2db2e9b9e96f1e7bb2/pkg/clusterversion/cockroach_versions.go#L536)):*
> ```Go
> 	// release, while on release branches it can be set to invalidVersionKey to
> 	// disable marking any versions as development versions.
> 	unstableVersionsAbove = V22_1
> ```

Ack. We'll update the stability period runbook accordingly, thanks.



_Code quote:_
```Go
	// release, while on release branches it can be set to invalidVersionKey to
	// disable marking any versions as development versions.
	unstableVersionsAbove = V22_1
```

**Comment by dt** [comment]
@ajwerner I'm going to leave this one to you to bors if you like it and approve it, or just close if you don't (which is also ok with me; I don't have my heart set on it).

**Review by ajwerner — APPROVED** [review]
:lgtm:

> Reviewed 4 of 4 files at r3, 9 of 12 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/86345#-:-NAK_j37C69OCJyI5hb0:bnfp4nl)* status: :shipit: complete! 2 of 0 LGTMs obtained (waiting on @bdarnell, @dt, and @stevendanna)

**Comment by celiala** [comment]
TFTR!

bors r=ajwerner

**Pull request merged**
